### PR TITLE
Added a whitelist column for whitelisting an ip and port together, and added support for the ip being a “*” 

### LINKED ISSUE #490: Have a whitelist column for whitelisting an ip and port togethr, and add support for the ip  being a “*”
Created by Alya Gomaa via monday.com integration. 🎉

### PR DESCRIPTION
**Pull request by Arsh-Khan**
## Fixes Issue
Closes #490 

## Changes proposed

For this issue, I have made modifications to the whitelist.conf file.
A new column has been introduced to facilitate support for whitelisting an IP and port together. 
Also added support for using wildcard (*) as a placeholder for IPs has been implemented. 
Additionally,  examples showcasing the usage of this feature have been included.

## Check List (Check all the applicable boxes) 

- [X] My code follows the code style of this project.
- [ ] My change requires changes to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed.
- [X] This PR does not contain plagiarized content.
- [X] The title of my pull request is a short description of the requested changes.

## Note to reviewers
If there are any further adjustments needed or if I've overlooked any modifications that should be made elsewhere, please don't hesitate to inform me. I'll address them promptly to ensure everything is in order. Just let me know, and I'll make the necessary updates as soon as possible!

### CODE REVIEW

**Comment by Arsh-Khan** [comment]
Hello @AlyaGomaa, Thank you for bringing this to my attention. I apologize for any confusion caused. I have just committed the file containing the handler code for the modifications to whitelist.conf.

**Comment by AlyaGomaa** [comment]
Hello @Arsh-Khan 

The integration tests are failing in your PR, can i ask you to make sure all unit tests and integration tests pass before i start reviewing your PR? these tests should pass if your code didn't break another functionality in Slips. so you should always run them before a PR.

If you don't know how to do this, check the contributing section in the docs.

Thanks!

**Comment by Arsh-Khan** [comment]
Hello @AlyaGomaa,
I've re-run all the tests and all checks have passed.
I'll make sure to run these tests before submitting any future PRs.
Please proceed with reviewing my PR, and let me know if you have any further questions or concerns.
Thank you

File: slips_files/core/helpers/whitelist.py [hunk 1]
```diff
@@ -552,6 +553,31 @@ def read_whitelist(self):
                                 # first time seeing this org
                                 whitelisted_orgs[data] = whitelist_line_info
 
+                        elif "whitelist" in type_:
```
Comment (AlyaGomaa) [inline slips_files/core/helpers/whitelist.py:556]: based on the example you put of the wildcard in the whitelist.conf. this condition is never triggered.

File: config/whitelist.conf [hunk 2]
```diff
@@ -42,6 +45,10 @@
 #organization,google,both,both
 #organization,apple,both,both
 #organization,twitter,both,both
+#ip,94.23.253.*,both,alerts,*
```
Comment (AlyaGomaa) [inline config/whitelist.conf:48]: This was never tested. and causes errors.

**Review by AlyaGomaa — COMMENTED** [review]
Hello, i see your code was never tested, and some of it will never be triggered. and it throws errors when running slips. i left you some comments in the review, i'm sorry i will have to close this PR.
In order to review it again, please provide a screenshot of the tests passing and a simple example of how you tested your code.
Thanks!

**Comment by Arsh-Khan** [comment]
Hello @AlyaGomaa, 

Updated the examples in whitelist.conf

The changes made in the read_whitelist function are:
If the data value is wildcard '*', it whitelists all IP addresses with the specified from, what_to_ignore, and port values.
If a port value is provided, it stores the IP address and port combination as the key in the whitelisted_IPs dictionary.

The changes made in the is_ip_whitelisted function are:

If the IP and port combination is found in the whitelisted_ip_ports dictionary, it checks the whitelisting direction and what to ignore (alerts or flows) to determine if the IP-port combination should be whitelisted.
If the IP is found in the whitelisted_ips dictionary, it follows the same logic as before to check if the IP should be whitelisted.

**Comment by AlyaGomaa** [comment]
Hello @Arsh-Khan 
Sure i can open the PR again, i will need you to run the tests locally and post a screenshot of the tests passing in the PR update, and tell me how you tested your PR with steps and how you made sure the changes you made are taking effect.


Once i check the tests and the steps you give me, i will re-open the PR and run the CI in github and if all goes well i'll merge it.

Thanks for you time and contribution. we appreciate it.

**Comment by Arsh-Khan** [comment]
Following changes were made:

1. Updating the Whitelist Configuration Examples:
The whitelist.conf file format was updated to include a new column for the whitelisted port.
The updated format is as follows:
```ip,[IP_ADDRESS],[BOTH|IN|OUT],[ALERTS|NONE],[PORT]```
Example entries:
```
ip,*,both,alerts,* - Whitelist all IP addresses and ports
ip,*,both,alerts - Whitelist all IP addresses, but not ports
ip,192.168.1.100,both,alerts,* - Whitelist IP 192.168.1.100 for all ports
```
2. Updating the read_whitelist Function:
The ```read_whitelist``` function was updated to parse the new whitelist config format, including the port column.
The function now reads each line of the whitelist.conf file, splits it by the delimiter, and extracts the necessary information (IP address, direction, action, and port).
The parsed data is then stored to be used in the subsequent functions.

3. Updating the parse_whitelist Function:
The ```parse_whitelist``` function was updated to handle the new whitelist configuration format.
The function now checks if the IP address is a ```wildcard (*)``` and if the port is a ```wildcard (*)```.
If the IP address is a wildcard, the function sets the whitelisted IP address to ```*``` to indicate that all IP addresses should be whitelisted.
If the port is a wildcard, the function sets the whitelisted port to ```*``` to indicate that all ports should be whitelisted.
The function then stores the parsed data (IP address, direction, action, and port).

4. Updating the is_ip_whitelisted Function:
The ```is_ip_whitelisted``` function was updated to handle the new whitelist configuration format.
The function now checks if the IP address and port are whitelisted by iterating through the parsed whitelist data.
If the IP address is a wildcard (```*``) or matches the provided IP address, and the port is a wildcard (```*```) or matches the provided port, the function returns ```True``` to indicate that the IP and port combination is whitelisted.
Otherwise, the function returns ```False```.

Here is a screenshot of the tests passing locally
Tests Results

**Comment by Arsh-Khan** [comment]
Hello @AlyaGomaa,
Please review my PR and approval to open the PR back so to proceed with running tests on the code in github.

**Pull request closed without merging**